### PR TITLE
Do not log the content of pod manifest if parsing fails.

### LINKED ISSUE #46493: Kubelet should only accept text/plain pod config file
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**Note:** Please file issues for subcomponents under the appropriate repo

| Component | Repo                                                               |
| --------- | ------------------------------------------------------------------ |
| kubectl   | [kubernetes/kubectl](https://github.com/kubernetes/kubectl/issues/new) |
| kubeadm   | [kubernetes/kubeadm](https://github.com/kubernetes/kubeadm/issues/new) |

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Kubernetes version** (use `kubectl version`):
HEAD

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
I start kubelet with flag `--pod-manifest-path` and want to put a pod config file there. But accidentally I copy a big binary file there, then the shell goes crazy! It goes on printing messy codes on and on because kubelet is trying to log the file content.

**What you expected to happen**:
We shouldn't log the config file content when it's invalid. Don't print messy codes and tell me the config file is invalid.

**How to reproduce it** (as minimally and precisely as possible):
See what happened.

**Anything else we need to know**:
I think we should
- just accept text/plain config file
- not log config file content when it's invalid, just log an error is enough.

Will send a PR to fix it.

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
- ~~only accepts text/plain config file~~
- ~~not log config file content when it's invalid~~

Do not log the content of pod manifest if parsing fails.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #46493

**Special notes for your reviewer**:
/cc @yujuhong 

@sig-node-reviewers

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/config/file.go [hunk 1]
```diff
@@ -200,3 +201,21 @@ func (s *sourceFile) replaceStore(pods ...*v1.Pod) (err error) {
...
 	return s.store.Replace(objs, "")
 }
+
+func checkPlainTextFile(file string) (bool, error) {
+	out, err := exec.Command("file", "--mime-type", file).CombinedOutput()
+	if err != nil {
+		return false, fmt.Errorf("failed to exec file command: %v", err)
+	}
+
+	lines := strings.Split(string(out), ": ")
```
Comment (dixudx) [inline pkg/kubelet/config/file.go]: I don't think this will work all the time. As you know, in linux, all the characters are valid except `/`

File: pkg/kubelet/config/file.go [hunk 2]
```diff
@@ -200,3 +201,21 @@ func (s *sourceFile) replaceStore(pods ...*v1.Pod) (err error) {
 	}
 	return s.store.Replace(objs, "")
 }
+
+func checkPlainTextFile(file string) (bool, error) {
+	out, err := exec.Command("file", "--mime-type", file).CombinedOutput()
```
Comment (dixudx) [inline pkg/kubelet/config/file.go]: I suggest using `out, err := exec.Command("mimetype", "--output-format=%m", file).CombinedOutput()` instead to avoid extra output split and analysis.

**Review by dixudx — COMMENTED** [review]
Need a better way to identify the content type.

**Comment by mtaufen** [comment]
If you're worried about too much spew from accidentally dropping a massive binary in the manifest dir (#46493), why not just limit the output to an arbitrary number of bytes in this case? MIME types aren't strongly guaranteed and I'd be worried not everyone has `text/plain` on these manifests.

File: pkg/kubelet/config/file.go (hunk 1 above)
Comment (xiangpengzhao) [inline pkg/kubelet/config/file.go]: I test it on the evn, it works well. Could you clarify the case when it won't wok? I might miss something.

File: pkg/kubelet/config/file.go (hunk 2 above)
Comment (xiangpengzhao) [inline pkg/kubelet/config/file.go]: Thanks for the suggestion! But most of hosts might not have `mimetype` by default, so the check will fail always.

Comment (dixudx) [inline pkg/kubelet/config/file.go]: @xiangpengzhao I think `out, err := exec.Command("file", "-b", "--mime-type", file).CombinedOutput()` will work.

**Comment by xiangpengzhao** [comment]
Thanks @mtaufen ! Restricting users to only use `text/plain` manifest is not an ideal way, as per the concern you mentioned. So I would like to remove the check and only keep the below modification.

```
 -	return pod, fmt.Errorf("%v: read '%v', but couldn't parse as pod(%v).\n",
 -		filename, string(data), podErr)
 +	return pod, fmt.Errorf("%v: couldn't parse as pod(%v), please check config file.\n", filename, podErr)
```

You also mentioned limiting the output to an arbitrary number of bytes. I don't think we need to output any content of the manifest in log. Just log the parsing error would be enough. Users can view their manifest by editor. WDYT?

File: pkg/kubelet/config/file.go (hunk 1 above)
Comment (dixudx) [inline pkg/kubelet/config/file.go]: If the manifest filename contains ": ", then the output string will have at least two ": ".

for example,

```shell
$ touch "file1: file1"
$ echo "test" > "file1: file1"
$ file --mime-type "file1: file1"
file1: file1: text/plain
```
 The length of the split result will be 3, not 2. However this should be considered valid input.

To avoid this circumstance, we just use brief mode not to prepend filenames to the output lines.

```shell
$ file -b --mime-type "file1: file1"
text/plain
```

File: pkg/kubelet/config/file.go (hunk 2 above)
Comment (xiangpengzhao) [inline pkg/kubelet/config/file.go]: Thanks @dixudx for your example and suggestion! That's Great! 

However, as per @mtaufen 's comment, restricting users to provide `text/plain` manifest might not be okay.

**Comment by mtaufen** [comment]
Yes, just logging the parse error should be enough.

**Comment by xiangpengzhao** [comment]
@mtaufen comment addressed. PTAL.

File: pkg/kubelet/config/file.go [hunk 3]
```diff
@@ -189,8 +189,7 @@ func (s *sourceFile) extractFromFile(filename string) (pod *v1.Pod, err error) {
 		return pod, nil
 	}
 
-	return pod, fmt.Errorf("%v: read '%v', but couldn't parse as pod(%v).\n",
-		filename, string(data), podErr)
+	return pod, fmt.Errorf("%v: couldn't parse as pod(%v), please check config file.\n", filename, podErr)
```
Comment (mtaufen) [inline pkg/kubelet/config/file.go]: nit: s/config file/manifest file
otherwise this looks good to me

Comment (xiangpengzhao) [inline pkg/kubelet/config/file.go]: In other places of file.go, the log uses `config file`, so I' prefer to use `config file` here also. Michael, WDYT?

L145: `glog.V(1).Infof("Can't process config file %q: %v", path, err)`
L157: `glog.V(3).Infof("Reading config file %q", filename)`

Comment (mtaufen) [inline pkg/kubelet/config/file.go]: That's fair. Fine to leave as-is in this PR then. 
Generally speaking, though, `config` is too ambiguous. It would be worth another PR to add clarity to these log messages IMO.

**Comment by mtaufen** [comment]
@k8s-bot ok to test

**Comment by falenn** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (183/150 (122%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 5470 characters long; only the first
------ 5000 or so are included here.
Received: from o9.sgmail.github.com ([167.89.101.2]:7529)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-5c7e-dev=imwiz.com@sgmail.github.com>)
	id 1dGU99-000R3K-SL
	for dev@imwiz.com; Thu, 01 Jun 2017 11:44:33 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=AckvQtRTCCkTpa2t3cPJ1WrM6B0=; b=ns8LCymgYdcqfFJt
	xplt0Ieo86hd47VcOdcofTv4z49UerMEI+emkoB3qYEf27bU5qb6KObwD+syLtM8
	BXUUJYpIS64Z1fHcPClWht2TLQfEF1u9HX4iXIII4Iydcrwrtv/vzP4UoTQ2kV0Q
	CDtjz99Mx0CuL2fjMEYIQJnUPHo=
Received: by filter1190p1mdw1.sendgrid.net with SMTP id filter1190p1mdw1-28803-59305274-6B
        2017-06-01 17:44:20.859349843 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0002p1iad1.sendgrid.net (SG) with ESMTP id ySAQ6MTFRgOqqaV5aHhJQw
	for <dev@imwiz.com>; Thu, 01 Jun 2017 17:44:20.831 +0000 (UTC)
Date: Thu, 01 Jun 2017 10:44:20 -0700

File: pkg/kubelet/config/file.go (hunk 3 above)
Comment (xiangpengzhao) [inline pkg/kubelet/config/file.go]: Thanks @mtaufen ! Will send a new PR after this merged.

**Comment by xiangpengzhao** [comment]
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this
